### ISSUE TITLE
Strange Error with Array definition for PDO options

### ISSUE THREAD

**Issue by a-sassermann** [PHP-DI/PHP-DI#640]
Hi all,
I am stuck on a problem where I wanted to grab the PDO Options from the di container.
Strangely, it fails with

Uncaught PDOException: PDO::__construct(): SQLSTATE[HY000]: General error: attribute value must be an integer

when using the container definition whereas when using the same array directly in PDO creation, it works. 

I can't  find any difference at the option array using print_r, vardump or within the IDE debugging.

Trying it on various boxes with different php versions seems as if this problem starts with php 7.1.

I have no clue what's happening,  test code follows, using php-di 6.0.5:

```
<?php

include_once 'vendor/autoload.php';

$containerBuilder = new \DI\ContainerBuilder();
$containerBuilder->addDefinitions([
    'db.climate.options' => [
        \PDO::ATTR_ERRMODE => \PDO::ERRMODE_EXCEPTION,
        \PDO::ATTR_DEFAULT_FETCH_MODE => \PDO::FETCH_ASSOC]
    ]
]);
$container = $containerBuilder->build();

// setting $o directly, test ok
$o1 = [
    \PDO::ATTR_ERRMODE => \PDO::ERRMODE_EXCEPTION,
    \PDO::ATTR_DEFAULT_FETCH_MODE => \PDO::FETCH_ASSOC
];
print_r($o1);
var_dump($o1);

$p1 = new \PDO('sqlite:test1.sq3', null, null, $o1);
unset ($p1);

// usind phpdi, test fails
$o2 = $container->get('db.climate.options');
print_r($o2);
var_dump($o2);
$p2 = new \PDO('sqlite:test2.sq3', null, null, $o2);
unset($p2);
```

```
php pdotest.php
Array
(
    [3] => 2
    [19] => 2
)
/home/user/PhpstormProjects/onion-climate-logger/pdotest.php:19:
array(2) {
  [3] =>
  int(2)
  [19] =>
  int(2)
}
Array
(
    [3] => 2
    [19] => 2
)
/home/user/PhpstormProjects/onion-climate-logger/pdotest.php:27:
array(2) {
  [3] =>
  int(2)
  [19] =>
  int(2)
}
PHP Fatal error:  Uncaught PDOException: PDO::__construct(): SQLSTATE[HY000]: General error: attribute value must be an integer in /home/user/PhpstormProjects/onion-climate-logger/pdotest.php:28
Stack trace:
#0 /home/user/PhpstormProjects/onion-climate-logger/pdotest.php(28): PDO->__construct('sqlite:test2.sq...', NULL, NULL, Array)
#1 {main}
  thrown in /home/user/PhpstormProjects/onion-climate-logger/pdotest.php on line 28

Process finished with exit code 255
```

Using: 
- PHP 7.2.12 (cli) (built: Nov  6 2018 15:07:37) ( NTS )  **Fails with Exception**
[PHP Modules]
Core
ctype
curl
date
dom
fileinfo
filter
gd
hash
intl
json
libxml
mbstring
mysqlnd
openssl
pcntl
pcre
PDO
pdo_mysql
pdo_sqlite
Phar
posix
readline
Reflection
session
SimpleXML
SPL
standard
tokenizer
xdebug
xml
xmlreader
xmlwriter
zip
zlib

[Zend Modules]
Xdebug

- PHP 7.1.1 (cli) ( NTS )   **Fails with Exception**
[PHP Modules]
Core
ctype
curl
date
filter
json
libxml
mbstring
openssl
pcre
PDO
pdo_sqlite
posix
Reflection
session
SimpleXML
SPL
sqlite3
standard
tokenizer
xml
zip
zlib

- PHP 7.0.30-0+deb9u1 (cli) (built: Jun 14 2018 13:50:25) ( NTS ) **Finished without Error**
PHP Modules]
apc
apcu
bz2
calendar
Core
ctype
curl
date
dom
exif
fileinfo
filter
ftp
gd
gettext
hash
iconv
intl
json
libxml
mbstring
mysqli
mysqlnd
openssl
pcntl
pcre
PDO
pdo_mysql
pdo_sqlite
Phar
posix
readline
Reflection
session
shmop
SimpleXML
sockets
SPL
sqlite3
ssh2
standard
sysvmsg
sysvsem
sysvshm
tokenizer
wddx
xdebug
xml
xmlreader
xmlwriter
xsl
Zend OPcache
zip
zlib

[Zend Modules]
Xdebug
Zend OPcache

**Comment by mnapoli** [comment]
Wow this is indeed very weird. And thanks for the code example.

I ran it again and get the same result as you (error) with PHP 7.2. I did have to fix line 9 because it had an extra `]` at the end of the line (but once fixed I get the same error as you).

The arrays seem to be exactly the same. I checked the array keys too and they are both integers too…

**Comment by mnapoli** [comment]
This sounds similar: https://github.com/Level-2/Dice/issues/95

**Comment by a-sassermann** [comment]
Based on your link to dice I did some more investigation. 
For me it seems to end up in some php internal structure corruption happening when doing the array_walk_recursive in ArrayResolver.php at line 45. Maybe you have a chance to avoid this construction.
Interestingly though it is not caused by the reference parameter &value, as it was described at the dice issue. Simply using array_walk_recursive is enough. And on top of it, you can fix it by doing a type check on the affected array values. I think I'm going file this at the php bugtracker. 

Just in case somebody is interested, here is the latest testcode, again tested with php7.0, 7.1 and 7.2 where 7.0 is without error:

```
<?php

function test($mode){
    $o = [
        \PDO::ATTR_ERRMODE => \PDO::ERRMODE_EXCEPTION,
        \PDO::ATTR_DEFAULT_FETCH_MODE => \PDO::FETCH_ASSOC
    ];

    var_dump($o);
    echo is_integer($o[\PDO::ATTR_ERRMODE ]).PHP_EOL;
    echo is_integer($o[\PDO::ATTR_DEFAULT_FETCH_MODE ]).PHP_EOL;


    if($mode != 'PDO'){
        // simply by having a walk across the array seems to destroy an internal type indicator
        array_walk_recursive($o, function ($value, $key) {
            //important stuff here
        });
        // but is not visible to the user
        var_dump($o);
    }

    if($mode == 'FIX-ARRAY-BY-CHECKING'){
        // just check the value types and it gets fixed.
        echo is_integer($o[\PDO::ATTR_ERRMODE ]).PHP_EOL;
        echo is_integer($o[\PDO::ATTR_DEFAULT_FETCH_MODE ]).PHP_EOL;
        // again no visible change
        var_dump($o);
    }

    $p = new \PDO('sqlite:test.sq3', null, null, $o);
    unset($p);
}
echo '**** Test 1 ***' . PHP_EOL;
echo 'Test PDO works as expected' . PHP_EOL;
echo '**** ****** ***' . PHP_EOL . PHP_EOL;
try{
    test('PDO');
}catch(\Exception $exception){
    echo $exception->getMessage();
}

echo PHP_EOL. PHP_EOL .'**** Test 2 ***' . PHP_EOL;
echo 'Test process PDO options with array_walk_recursive' . PHP_EOL;
echo '**** ****** ***' . PHP_EOL . PHP_EOL;

try{
    test('ARRAY_WALK');
}catch(\Exception $exception){
    echo $exception->getMessage();
}

echo PHP_EOL. PHP_EOL .'**** Test 3 ***' . PHP_EOL;
echo 'Test justify internal type by checking type of array values' . PHP_EOL;
echo '**** ****** ***' . PHP_EOL . PHP_EOL;
try{
    test('FIX-ARRAY-BY-CHECKING');
}catch(\Exception $exception){
    echo $exception->getMessage();
}

```

**Comment by mnapoli** [comment]
Wow nice job investigating that. If you have a link to the bug in PHP's bugtracker I'm interested!

**Comment by a-sassermann** [comment]
Thanks :-)
Bugreport is here: https://bugs.php.net/bug.php?id=77273
I'm curious about the result...

**Comment by a-sassermann** [comment]
Bug has been confirmed by php.
Fix is in Versions 7.3.2 and 7.2.15

**Comment by mnapoli** [comment]
Awesome, thank you for the followup @a-sassermann!
